dask/order.py

```diff
@@ -202,125 +193,212 @@ def dependencies_key(x):
+                                    now_keys.sort(reverse=True)
+                                for key in now_keys:
+                                    pool = dep_pools[key]
+                                    if 1 < len(pool) < 100:
+                                        pool.sort(key=dependents_key, reverse=True)
+                                    inner_stacks_extend([dep] for dep in pool)
+                                inner_stack = inner_stacks_pop()
                         else:
-                            # Our next starting point will be "seeded" by a completed task in a
-                            # FIFO manner.  When our DFS with `inner_stack` is finished--which
-                            # means we computed our tactical goal--we will choose our next starting
-                            # point from the dependents of completed tasks.  However, it is too
-                            # expensive to consider all dependents of all completed tasks, so we
-                            # consider the dependents of tasks in the order they complete.
-                            outer_stack_seeds_append(item)
-
-            while not outer_stack and outer_stack_seeds_index < len(outer_stack_seeds):
-                # We wait for as long as possible to consider dependents of completed nodes:
-                # 1. some may have already completed, so waiting lets us sort fewer items, and
-                # 2. sorting via `dependents_key` depends on `num_needed`, which is dynamic.
-                deps = set_difference(
-                    dependents[outer_stack_seeds[outer_stack_seeds_index]], result
-                )
-                if deps:
-                    if 1 < len(deps) < 1000:
-                        outer_stack_extend(
-                            sorted(deps, key=dependents_key, reverse=True)
-                        )
-                    else:
-                        outer_stack_extend(deps)
-                outer_stack_seeds_index += 1
+                            min_key = min(dep_pools)
+                            min_pool = dep_pools[min_key]
+                            if 1 < len(min_pool) < 100:
+                                dep = min(min_pool, key=dependents_key)
+                            else:
+                                dep = min_pool.pop()
+                                if not min_pool:
+                                    del dep_pools[min_key]
+                            inner_stack = [dep]
+                            for key, vals in dep_pools.items():
+                                if key < item_key:
+                                    if key in next_nodes:
+                                        next_nodes[key].append(vals)
+                                    else:
+                                        next_nodes[key] = [vals]
+                                elif key in later_nodes:
+                                    later_nodes[key].append(vals)
```

Comment:
Appears to be untested.

Reply:
Now covered after switching to `defaultdict(list)`